File: androidTestApp/src/main/java/edu/gatech/cc/cellwatch/androidtestapp/designsystem/Theme.kt
package edu.gatech.cc.cellwatch.androidtestapp.designsystem

import android.content.Context
import android.graphics.Color
import android.util.TypedValue

/**
 * Visual baseline for product screens. Mirror of the iOS `Theme`.
 *
 * Both platforms carry the same palette, the same 4-based spacing scale and the
 * same template so the two apps stay recognisably one product. The palette is
 * frozenApp's `res/values/colors.xml`; the scale is what its 21 layouts
 * actually used - 16dp 58 times, 8dp 47, 4dp 17, 12dp 15, 24dp 7 - with the
 * off-scale values (2, 5, 6, 10dp) being the drift this replaces.
 *
 * Values are dp, converted at use. Nothing here is a raw pixel.
 */
object Theme {

    object Palette {
        const val BLUE = 0xFF07416B.toInt()
        const val BLUE_LIGHT = 0xFF5E8BAB.toInt()
        const val GREEN = 0xFF1E5638.toInt()
        const val GREEN_LIGHT = 0xFFC8E3CC.toInt()
        const val GREY = 0xFF777777.toInt()
        const val GREY_DARK = 0xFF464646.toInt()
        const val GREY_EXTRA_LIGHT = 0xFFDFDFDF.toInt()
        const val GREY_ULTRA_LIGHT = 0xFFF2F2F2.toInt()
        const val ORANGE = 0xFFB44D0D.toInt()

        // Roles. Screens use these, not raw hues.
        const val PRIMARY = BLUE
        const val ON_PRIMARY = Color.WHITE
        const val SUCCESS = GREEN
        const val WARNING = ORANGE
        const val SURFACE = Color.WHITE
        const val BACKGROUND = GREY_ULTRA_LIGHT
        const val BORDER = GREY_EXTRA_LIGHT
        const val TEXT_PRIMARY = 0xFF1A1A1A.toInt()
        const val TEXT_SECONDARY = GREY_DARK
        const val TEXT_MUTED = GREY
    }

    /** dp. Names, not numbers, at call sites. */
    object Space {
        const val XS = 4
        const val S = 8
        const val M = 12
        const val L = 16
        const val XL = 24
        const val XXL = 32
    }

    object Radius {
        const val CONTROL = 10
        const val CARD = 12
    }

    /** sp. Matches the iOS type ramp. */
    object TextSize {
        const val TITLE = 24f
        const val HEADING = 18f
        const val BODY = 16f
        const val CAPTION = 13f
        const val METRIC = 17f
    }

    /** Material's minimum touch target, and iOS's. */
    const val MIN_TAP_TARGET_DP = 48

    fun Context.dp(value: Int): Int = TypedValue.applyDimension(
        TypedValue.COMPLEX_UNIT_DIP,
        value.toFloat(),
        resources.displayMetrics,
    ).toInt()
}

File: iosTestApp/App/DesignSystem/Theme.swift
import UIKit

/// Visual baseline for product screens.
///
/// Derived from `frozenApp`, not invented: the palette is its `colors.xml`, and
/// the spacing scale is what its 21 layouts actually used — 16dp appears 58
/// times, 8dp 47, 4dp 17, 12dp 15, 24dp 7. Values outside that scale (2, 5, 6,
/// 10dp) are the ad-hoc drift this replaces; frozenApp's own `dimens.xml` was
/// empty, so spacing was decided per layout.
///
/// Screens use these tokens. A screen that needs a value not here changes the
/// scale deliberately rather than hard-coding a one-off.
enum Theme {

    /// CellWatch palette, from frozenApp `res/values/colors.xml`.
    enum Color {
        static let blue = UIColor(hex: 0x07416B)
        static let blueLight = UIColor(hex: 0x5E8BAB)
        static let green = UIColor(hex: 0x1E5638)
        static let greenLight = UIColor(hex: 0xC8E3CC)
        static let greenDark = UIColor(hex: 0x003618)
        static let grey = UIColor(hex: 0x777777)
        static let greyDark = UIColor(hex: 0x464646)
        static let greyLight = UIColor(hex: 0xBABABA)
        static let greyExtraLight = UIColor(hex: 0xDFDFDF)
        static let greyUltraLight = UIColor(hex: 0xF2F2F2)
        static let orange = UIColor(hex: 0xB44D0D)

        // Roles. Screens reference these, not the raw palette, so a palette
        // change lands in one place.
        static let primary = blue
        static let onPrimary = UIColor.white
        static let success = green
        static let warning = orange
        static let surface = UIColor.white
        static let background = greyUltraLight
        static let border = greyExtraLight
        static let textPrimary = UIColor(hex: 0x1A1A1A)
        static let textSecondary = greyDark
        static let textMuted = grey
    }

    /// 4-based scale. Names, not numbers, at call sites.
    enum Space {
        static let xs: CGFloat = 4
        static let s: CGFloat = 8
        static let m: CGFloat = 12
        static let l: CGFloat = 16
        static let xl: CGFloat = 24
        static let xxl: CGFloat = 32
    }

    enum Radius {
        static let control: CGFloat = 10
        static let card: CGFloat = 12
    }

    /// Dynamic Type throughout: these scale with the user's text size setting,
    /// which fixed-point fonts would not.
    enum Font {
        static let title = UIFont.preferredFont(forTextStyle: .title2).semibold()
        static let heading = UIFont.preferredFont(forTextStyle: .headline)
        static let body = UIFont.preferredFont(forTextStyle: .body)
        static let caption = UIFont.preferredFont(forTextStyle: .footnote)
        static let metric = UIFont.monospacedDigitSystemFont(ofSize: 17, weight: .semibold)
    }

    /// Minimum tap target. Apple's guidance is 44pt; several existing harness
    /// buttons are smaller.
    static let minimumTapTarget: CGFloat = 44
}

extension UIColor {
    convenience init(hex: UInt32) {
        self.init(
            red: CGFloat((hex >> 16) & 0xFF) / 255.0,
            green: CGFloat((hex >> 8) & 0xFF) / 255.0,
            blue: CGFloat(hex & 0xFF) / 255.0,
            alpha: 1.0
        )
    }
}

extension UIFont {
    func semibold() -> UIFont {
        guard let descriptor = fontDescriptor.withSymbolicTraits(.traitBold) else { return self }
        return UIFont(descriptor: descriptor, size: 0)
    }
}
